crates/hir/src/source_analyzer.rs

```diff
@@ -224,14 +224,18 @@ impl SourceAnalyzer {
     ) -> Option<PathResolution> {
         if let Some(path_expr) = path.syntax().parent().and_then(ast::PathExpr::cast) {
             let expr_id = self.expr_id(db, &path_expr.into())?;
-            if let Some(assoc) = self.infer.as_ref()?.assoc_resolutions_for_expr(expr_id) {
+            let infer = self.infer.as_ref()?;
+            if let Some(assoc) = infer.assoc_resolutions_for_expr(expr_id) {
                 return Some(PathResolution::AssocItem(assoc.into()));
             }
             if let Some(VariantId::EnumVariantId(variant)) =
-                self.infer.as_ref()?.variant_resolution_for_expr(expr_id)
+                infer.variant_resolution_for_expr(expr_id)
             {
                 return Some(PathResolution::Def(ModuleDef::Variant(variant.into())));
             }
+            if let Some(func) = infer[expr_id].as_fn_def() {
+                return Some(PathResolution::Def(ModuleDef::Function(func.into())));
+            }
```

Comment:
I guess this is a  good enough fix, but it still is not entirey correct. I thing this'll fail for

```
fn foo() {}
fn main() {
    let bar = foo;
    bar() // <- this should be a local, rather than a function. 
}
```

Reply:
Hmm right, that fails this example then 😕

Reply:
I wonder though, why can't we just make this resolve to a variable proprely? Iike, if ther's no shadowing, we do resolve the right thing. So the code is already there, we just need to tweak order of things a bit?

Reply:
Hm, that's a good point, since we are inside a `PathExpr` here it has to be a value no matter what right? I guess swapping the order in that case should fix this properly then? I'll test that out.